// @flow-js/garmin-connect/dist/garmin/workout-builder/duration.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeartRateDuration = exports.CaloriesDuration = exports.LapPressDuration = exports.DistanceDuration = exports.TimeDuration = exports.NoDuration = exports.Duration = void 0;
var Duration = /** @class */ (function () {
    function Duration() {
    }
    return Duration;
}());
exports.Duration = Duration;
var NoDuration = /** @class */ (function (_super) {
    __extends(NoDuration, _super);
    function NoDuration() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NoDuration.prototype.build = function () {
        throw new Error('Please provide an duration for this step (Other than NoDuration).');
    };
    return NoDuration;
}(Duration));
exports.NoDuration = NoDuration;
var TimeDuration = /** @class */ (function (_super) {
    __extends(TimeDuration, _super);
    function TimeDuration(durationInSeconds) {
        var _this = _super.call(this) || this;
        _this.durationInSeconds = durationInSeconds;
        return _this;
    }
    TimeDuration.hhmmss = function (h, m, s) {
        return new TimeDuration(h * 3600 + m * 60 + s);
    };
    TimeDuration.fromHours = function (h) {
        return new TimeDuration(h * 3600);
    };
    TimeDuration.fromMinutes = function (m) {
        return new TimeDuration(m * 60);
    };
    TimeDuration.fromSeconds = function (s) {
        return new TimeDuration(s);
    };
    TimeDuration.prototype.build = function () {
        return {
            endCondition: {
                conditionTypeId: 2,
                conditionTypeKey: 'time',
                displayable: true,
                displayOrder: 1
            },
            endConditionCompare: null,
            endConditionValue: this.durationInSeconds,
            endConditionZone: null
        };
    };
    return TimeDuration;
}(Duration));
exports.TimeDuration = TimeDuration;
var DistanceDuration = /** @class */ (function (_super) {
    __extends(DistanceDuration, _super);
    function DistanceDuration(meters) {
        var _this = _super.call(this) || this;
        _this.meters = meters;
        return _this;
    }
    DistanceDuration.fromKilometers = function (km) {
        return new DistanceDuration(km * 1000);
    };
    DistanceDuration.fromMeters = function (m) {
        return new DistanceDuration(m);
    };
    DistanceDuration.prototype.build = function () {
        return {
            endCondition: {
                conditionTypeId: 3,
                conditionTypeKey: 'distance',
                displayOrder: 3,
                displayable: true
            },
            endConditionValue: this.meters,
            preferredEndConditionUnit: {
                unitKey: 'kilometer'
            },
            endConditionCompare: null
        };
    };
    return DistanceDuration;
}(Duration));
exports.DistanceDuration = DistanceDuration;
var LapPressDuration = /** @class */ (function (_super) {
    __extends(LapPressDuration, _super);
    function LapPressDuration() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LapPressDuration.prototype.build = function () {
        return {
            endCondition: {
                conditionTypeId: 1,
                conditionTypeKey: 'lap.button',
                displayOrder: 1,
                displayable: true
            },
            endConditionValue: null,
            preferredEndConditionUnit: null,
            endConditionCompare: null
        };
    };
    return LapPressDuration;
}(Duration));
exports.LapPressDuration = LapPressDuration;
var CaloriesDuration = /** @class */ (function (_super) {
    __extends(CaloriesDuration, _super);
    function CaloriesDuration(calories) {
        var _this = _super.call(this) || this;
        _this.calories = calories;
        return _this;
    }
    CaloriesDuration.prototype.build = function () {
        return {
            endCondition: {
                conditionTypeId: 4,
                conditionTypeKey: 'calories',
                displayOrder: 4,
                displayable: true
            },
            endConditionValue: this.calories
        };
    };
    return CaloriesDuration;
}(Duration));
exports.CaloriesDuration = CaloriesDuration;
var HeartRateDuration = /** @class */ (function (_super) {
    __extends(HeartRateDuration, _super);
    function HeartRateDuration(bpm, compare) {
        var _this = _super.call(this) || this;
        _this.bpm = bpm;
        _this.compare = compare;
        return _this;
    }
    HeartRateDuration.greaterThan = function (bpm) {
        return new HeartRateDuration(bpm, 'gt');
    };
    HeartRateDuration.lessThan = function (bpm) {
        return new HeartRateDuration(bpm, 'lt');
    };
    HeartRateDuration.prototype.build = function () {
        return {
            endCondition: {
                conditionTypeId: 6,
                conditionTypeKey: 'heart.rate',
                displayOrder: 6,
                displayable: true
            },
            endConditionValue: this.bpm,
            endConditionCompare: this.compare
        };
    };
    return HeartRateDuration;
}(Duration));
exports.HeartRateDuration = HeartRateDuration;
