// @flow-js/garmin-connect/dist/garmin/workout-builder/step.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepType = exports.Step = void 0;
var duration_1 = require("./duration");
var target_1 = require("./target");
var Step = /** @class */ (function () {
    function Step(stepType, duration, target, notes) {
        if (duration === void 0) { duration = new duration_1.NoDuration(); }
        if (target === void 0) { target = new target_1.NoTarget(); }
        if (notes === void 0) { notes = ''; }
        this.stepType = stepType;
        this.duration = duration;
        this.target = target;
        this.notes = notes;
    }
    Step.prototype.build = function (index) {
        return __assign(__assign(__assign({ category: null, childStepId: null, description: this.notes, equipmentType: {
                displayOrder: null,
                equipmentTypeId: null,
                equipmentTypeKey: null
            }, exerciseName: null, preferredEndConditionUnit: null, providerExerciseSourceId: null, secondaryTargetType: null, secondaryTargetValueOne: null, secondaryTargetValueTwo: null, secondaryTargetValueUnit: null, secondaryZoneNumber: null, stepAudioNote: null, stepId: index, stepOrder: index, strokeType: {}, targetValueOne: null, targetValueTwo: null, targetValueUnit: null, type: 'ExecutableStepDTO', weightUnit: null, weightValue: null, workoutProvider: null, zoneNumber: null }, this.stepType.build()), this.duration.build()), this.target.build());
    };
    return Step;
}());
exports.Step = Step;
var StepType = /** @class */ (function () {
    function StepType(type, stepId) {
        this.type = type;
        this.stepId = stepId;
    }
    StepType.prototype.build = function () {
        return {
            stepType: {
                stepTypeId: this.stepId,
                stepTypeKey: this.type,
                displayOrder: this.stepId
            }
        };
    };
    StepType.WarmUp = new StepType('warmup', 1);
    StepType.Run = new StepType('interval', 3);
    StepType.Recovery = new StepType('recovery', 4);
    StepType.Rest = new StepType('rest', 5);
    StepType.Cooldown = new StepType('cooldown', 2);
    StepType.Other = new StepType('other', 7);
    return StepType;
}());
exports.StepType = StepType;
