// @flow-js/garmin-connect/dist/garmin/workout-builder/target.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PowerZone = exports.PowerZoneTarget = exports.HrmTarget = exports.HrmZoneTarget = exports.CadenceTarget = exports.PaceTarget = exports.NoTarget = exports.Target = void 0;
var Target = /** @class */ (function () {
    function Target() {
    }
    return Target;
}());
exports.Target = Target;
var NoTarget = /** @class */ (function (_super) {
    __extends(NoTarget, _super);
    function NoTarget() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NoTarget.prototype.build = function () {
        return {
            targetType: {
                workoutTargetTypeId: 1,
                workoutTargetTypeKey: 'no.target',
                displayOrder: 1
            }
        };
    };
    return NoTarget;
}(Target));
exports.NoTarget = NoTarget;
var PaceTarget = /** @class */ (function (_super) {
    __extends(PaceTarget, _super);
    /**
     * Create a pace target
     * @param minPace - in seconds per meter
     * @param maxPace - in seconds per meter
     */
    function PaceTarget(minPace, maxPace) {
        var _this = _super.call(this) || this;
        _this.minPace = minPace;
        _this.maxPace = maxPace;
        return _this;
    }
    PaceTarget.pace = function (paceMinutes, paceSeconds, margin) {
        if (margin === void 0) { margin = 0; }
        var minPace = paceMinutes * 60 + paceSeconds + margin;
        var maxPace = paceMinutes * 60 + paceSeconds - margin;
        return new PaceTarget(1000 / minPace, 1000 / maxPace);
    };
    PaceTarget.prototype.build = function () {
        return {
            targetType: {
                workoutTargetTypeId: 6,
                workoutTargetTypeKey: 'pace.zone',
                displayOrder: 6
            },
            targetValueOne: this.minPace,
            targetValueTwo: this.maxPace,
            targetValueUnit: null
        };
    };
    return PaceTarget;
}(Target));
exports.PaceTarget = PaceTarget;
var CadenceTarget = /** @class */ (function (_super) {
    __extends(CadenceTarget, _super);
    function CadenceTarget(minCadence, maxCadence) {
        var _this = _super.call(this) || this;
        _this.minCadence = minCadence;
        _this.maxCadence = maxCadence;
        return _this;
    }
    CadenceTarget.cadence = function (cadence, margin) {
        return new CadenceTarget(cadence - margin, cadence + margin);
    };
    CadenceTarget.prototype.build = function () {
        return {
            targetType: {
                workoutTargetTypeId: 3,
                workoutTargetTypeKey: 'cadence',
                displayOrder: 3
            },
            targetValueOne: this.minCadence,
            targetValueTwo: this.maxCadence,
            targetValueUnit: null
        };
    };
    return CadenceTarget;
}(Target));
exports.CadenceTarget = CadenceTarget;
var HrmZoneTarget = /** @class */ (function (_super) {
    __extends(HrmZoneTarget, _super);
    function HrmZoneTarget(hrmZone) {
        var _this = _super.call(this) || this;
        _this.hrmZone = hrmZone;
        return _this;
    }
    HrmZoneTarget.prototype.build = function () {
        return {
            targetType: {
                workoutTargetTypeId: 4,
                workoutTargetTypeKey: 'heart.rate.zone',
                displayOrder: 4
            },
            zoneNumber: this.hrmZone
        };
    };
    return HrmZoneTarget;
}(Target));
exports.HrmZoneTarget = HrmZoneTarget;
var HrmTarget = /** @class */ (function (_super) {
    __extends(HrmTarget, _super);
    function HrmTarget(minHrm, maxHrm) {
        var _this = _super.call(this) || this;
        _this.minHrm = minHrm;
        _this.maxHrm = maxHrm;
        return _this;
    }
    HrmTarget.hrm = function (hrm, margin) {
        return new HrmTarget(hrm - margin, hrm + margin);
    };
    HrmTarget.prototype.build = function () {
        return {
            targetType: {
                workoutTargetTypeId: 4,
                workoutTargetTypeKey: 'heart.rate.zone',
                displayOrder: 4
            },
            targetValueOne: this.minHrm,
            targetValueTwo: this.maxHrm,
            targetValueUnit: null
        };
    };
    return HrmTarget;
}(Target));
exports.HrmTarget = HrmTarget;
var PowerZoneTarget = /** @class */ (function (_super) {
    __extends(PowerZoneTarget, _super);
    function PowerZoneTarget(powerZone) {
        var _this = _super.call(this) || this;
        _this.powerZone = powerZone;
        return _this;
    }
    PowerZoneTarget.prototype.build = function () {
        return {
            targetType: {
                workoutTargetTypeId: 2,
                workoutTargetTypeKey: 'power.zone',
                displayOrder: 2
            },
            zoneNumber: this.powerZone
        };
    };
    return PowerZoneTarget;
}(Target));
exports.PowerZoneTarget = PowerZoneTarget;
var PowerZone = /** @class */ (function (_super) {
    __extends(PowerZone, _super);
    function PowerZone(minPower, maxPower) {
        var _this = _super.call(this) || this;
        _this.minPower = minPower;
        _this.maxPower = maxPower;
        return _this;
    }
    PowerZone.power = function (power, margin) {
        return new PowerZone(power - margin, power + margin);
    };
    PowerZone.prototype.build = function () {
        return {
            targetType: {
                workoutTargetTypeId: 2,
                workoutTargetTypeKey: 'power.zone',
                displayOrder: 2
            },
            targetValueOne: this.minPower,
            targetValueTwo: this.maxPower
        };
    };
    return PowerZone;
}(Target));
exports.PowerZone = PowerZone;
