// @flow-js/garmin-connect/dist/garmin/workout-builder/workout-builder.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkoutBuilder = void 0;
var WorkoutBuilder = /** @class */ (function () {
    function WorkoutBuilder(workoutType, workoutName, workoutDescription) {
        if (workoutDescription === void 0) { workoutDescription = ''; }
        this.workoutType = workoutType;
        this.workoutName = workoutName;
        this.workoutDescription = workoutDescription;
        this.steps = [];
    }
    WorkoutBuilder.prototype.addStep = function (step) {
        this.steps.push(step);
        return this;
    };
    WorkoutBuilder.prototype.build = function () {
        return __assign(__assign({}, this.workoutType.build()), { subSportType: null, workoutName: this.workoutName, description: this.workoutDescription, estimatedDistanceUnit: { unitKey: null }, workoutSegments: [
                __assign(__assign({ segmentOrder: 1 }, this.workoutType.build()), { workoutSteps: this.steps.map(function (step, index) {
                        return step.build(index + 1);
                    }) })
            ], avgTrainingSpeed: null, estimatedDurationInSecs: 0, estimatedDistanceInMeters: 0, estimateType: null, isWheelchair: false });
    };
    return WorkoutBuilder;
}());
exports.WorkoutBuilder = WorkoutBuilder;
