// @flow-js/garmin-connect/dist/garmin/common/GpxUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.courseRequestTemplate = void 0;
var types_1 = require("../types");
function courseRequestTemplate(activityType, courseName, geoPoints, coursePoints, privacy) {
    if (coursePoints === void 0) { coursePoints = []; }
    if (privacy === void 0) { privacy = types_1.CoursePrivacyRule.PRIVATE; }
    return {
        activityTypePk: activityType,
        hasTurnDetectionDisabled: false,
        geoPoints: geoPoints,
        courseLines: [],
        coursePoints: coursePoints,
        startPoint: geoPoints[0],
        elapsedSeconds: null,
        openStreetMap: false,
        coordinateSystem: 'WGS84',
        rulePK: privacy,
        courseName: courseName,
        matchedToSegments: false,
        includeLaps: false,
        hasPaceBand: false,
        hasPowerGuide: false,
        favorite: false,
        speedMeterPerSecond: null,
        sourceTypeId: 3
    };
}
exports.courseRequestTemplate = courseRequestTemplate;
