// @flow-js/garmin-connect/dist/garmin/types/gpx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoursePrivacyRule = exports.GpxActivityType = void 0;
var GpxActivityType;
(function (GpxActivityType) {
    GpxActivityType[GpxActivityType["Running"] = 1] = "Running";
    GpxActivityType[GpxActivityType["Running_Road"] = 1] = "Running_Road";
    GpxActivityType[GpxActivityType["Running_Trail"] = 1] = "Running_Trail";
    GpxActivityType[GpxActivityType["Cycling"] = 10] = "Cycling";
    GpxActivityType[GpxActivityType["Cycling_Road"] = 10] = "Cycling_Road";
    GpxActivityType[GpxActivityType["Cycling_Gravel"] = 143] = "Cycling_Gravel";
    GpxActivityType[GpxActivityType["Hiking"] = 3] = "Hiking";
    GpxActivityType[GpxActivityType["Other"] = 4] = "Other";
})(GpxActivityType = exports.GpxActivityType || (exports.GpxActivityType = {}));
var CoursePrivacyRule;
(function (CoursePrivacyRule) {
    CoursePrivacyRule[CoursePrivacyRule["PUBLIC"] = 1] = "PUBLIC";
    CoursePrivacyRule[CoursePrivacyRule["PRIVATE"] = 2] = "PRIVATE";
})(CoursePrivacyRule = exports.CoursePrivacyRule || (exports.CoursePrivacyRule = {}));
