// @flow-js/garmin-connect/dist/garmin/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoursePrivacyRule = exports.GpxActivityType = exports.ActivitySubType = exports.ActivityType = exports.UploadFileType = exports.ExportFileType = void 0;
var ExportFileType;
(function (ExportFileType) {
    ExportFileType["tcx"] = "tcx";
    ExportFileType["gpx"] = "gpx";
    ExportFileType["kml"] = "kml";
    ExportFileType["zip"] = "zip";
})(ExportFileType = exports.ExportFileType || (exports.ExportFileType = {}));
var UploadFileType;
(function (UploadFileType) {
    UploadFileType["tcx"] = "tcx";
    UploadFileType["gpx"] = "gpx";
    UploadFileType["fit"] = "fit";
})(UploadFileType = exports.UploadFileType || (exports.UploadFileType = {}));
// Export other types
var activity_1 = require("./activity");
Object.defineProperty(exports, "ActivityType", { enumerable: true, get: function () { return activity_1.ActivityType; } });
Object.defineProperty(exports, "ActivitySubType", { enumerable: true, get: function () { return activity_1.ActivitySubType; } });
var gpx_1 = require("./gpx");
Object.defineProperty(exports, "GpxActivityType", { enumerable: true, get: function () { return gpx_1.GpxActivityType; } });
Object.defineProperty(exports, "CoursePrivacyRule", { enumerable: true, get: function () { return gpx_1.CoursePrivacyRule; } });
